pkg/kubelet/kubelet.go

```diff
@@ -1573,26 +1573,130 @@ func (kl *Kubelet) syncLoop(updates <-chan PodUpdate, handler SyncHandler) {
 	}
 }
 
-// Updated the Kubelet's internal pods with those provided by the update.
+func isStaticPod(pod *api.BoundPod) bool {
+	if getPodSource(pod) != ApiserverSource {
+		return true
+	}
+	return false
+}
+
+func isMirrorPod(pod *api.BoundPod) bool {
+	if value, ok := pod.Annotations[ConfigMirrorAnnotationKey]; !ok {
+		return false
+	} else {
+		return value == MirrorType
+	}
+}
+
+func (kl *Kubelet) createMirrorPod(pod *api.BoundPod) error {
+	if kl.kubeClient == nil {
+		return nil
+	}
+	// Create a Pod from BoundPod; this will become redundant once we
+	// deprecate BoundPod completely.
+	// Note that some fields (e.g. creationTimestamp, UID, etc) will be
+	// overwritten by the APIserver.
+	podToCreate := api.Pod{}
+	if err := api.Scheme.Convert(pod, &podToCreate); err != nil {
+		glog.Errorf("Unable to interpret static bound pod as a pod: %v: %+v", err, pod)
+	}
+	// Indicate that the pod should be scheduled to the current node.
+	podToCreate.Spec.Host = kl.hostname
+	podToCreate.Annotations[ConfigMirrorAnnotationKey] = MirrorType
+
+	_, err := kl.kubeClient.Pods(NamespaceDefault).Create(&podToCreate)
+	if err != nil {
+		glog.Errorf("Failed creating a mirror pod %q: %#v", GetPodFullName(pod), err)
+		return err
+	}
+	return nil
+}
+
+func (kl *Kubelet) deleteMirrorPod(pod *api.BoundPod) error {
+	if kl.kubeClient == nil {
+		return nil
+	}
+	err := kl.kubeClient.Pods(NamespaceDefault).Delete(pod.Name)
+	if err != nil {
+		glog.Errorf("Failed deleting a mirror pod %q: %v", GetPodFullName(pod), err)
+		return err
+	}
+	return nil
+}
+
+// The pods from sources other than Apiserver are called static pods. In order
+// to monitor such pods from the Apiserver, this function creates (and deletes)
+// mirror API pods for the static pods. The function returns pods with static
+// pods replaced by the mirror pods.
+func (kl *Kubelet) handleStaticPods(pods []api.BoundPod) []api.BoundPod {
+	// If kubeClient is not set, we can't report static pods back to the
+	// API server, so skip the extra logic.
+	if kl.kubeClient == nil {
+		return pods
+	}
+	// Actually handle the static pods.
+	return kl.staticPodWorker(pods)
+}
+
+func (kl *Kubelet) staticPodWorker(pods []api.BoundPod) []api.BoundPod {
+	// Create a podFullName -> count map.
+	numPodsByName := make(map[string]int)
+	for _, pod := range pods {
+		name := GetPodFullName(&pod)
+		if num, ok := numPodsByName[name]; ok {
```

Comment:
nit: `numPodsByName[name]++` will do the right thing. It it exists it increments it and if it doesn't it will create it (with default value 0) and increment it.